provider/pkg/provider/command.go

```diff
@@ -153,9 +181,79 @@ func (c *command) run(ctx context.Context, command string, host *provider.HostCl
 		return "", "", "", err
 	}
 
+	if c.AssetPaths != nil {
+		assets, err := globAssets(cmd.Dir, *c.AssetPaths)
+		if err != nil {
+			return "", "", "", err
+		}
+		c.Assets = &assets
+	}
+
+	if c.ArchivePaths != nil {
+		archiveAssets := map[string]interface{}{}
+		assets, err := globAssets(cmd.Dir, *c.ArchivePaths)
+		if err != nil {
+			return "", "", "", err
+		}
+
+		for path, asset := range assets {
+			archiveAssets[path] = asset
+		}
+
+		archive, err := resource.NewAssetArchive(archiveAssets)
+		if err != nil {
+			return "", "", "", err
+		}
+		c.Archive = archive
+	}
+
 	return strings.TrimSuffix(stdoutbuf.String(), "\n"), strings.TrimSuffix(stderrbuf.String(), "\n"), id, nil
 }
 
+func globAssets(dir string, globs []string) (map[string]*resource.Asset, error) {
+	assets := map[string]*resource.Asset{}
+	compiledGlobs := make([]glob.Glob, len(globs))
+	isGlobExclude := make([]bool, len(globs))
+	for i, g := range globs {
+		isExclude := strings.HasPrefix(g, "!")
+		g = strings.TrimPrefix(g, "!")
+		compiled, err := glob.Compile(g, '/')
+		if err != nil {
+			return nil, err
+		}
+		compiledGlobs[i] = compiled
+		isGlobExclude[i] = isExclude
+	}
+
+	err := fs.WalkDir(os.DirFS(dir), ".", func(p string, d fs.DirEntry, err error) error {
```

Comment:
I run this on the node_modules in the example and it doesn't have much delay. Had a search for if there's a smarter globbing library available which would allow us to skip iteration of subfolders based on glob rules but nothing seems to be available in go right now. Wouldn't be keen to hand-code something here to optimise for if super-wildcard exclude patterns etc as likely to get it wrong. 

The optimisation path here would probably be to drop back to doing this outside of Pulumi or building into a separate directory without all the superfluous files.